core/src/test/scala/unit/kafka/server/FetchRequestTest.scala

```diff
@@ -169,6 +171,58 @@ class FetchRequestTest extends BaseRequestTest {
   }
 
   /**
+   * Tests that down-conversions dont leak memory. Large down conversions are triggered
+   * in the server where the client closes connection. If references to byte arrays are
+   * not released this will OOM.
+   */
+  @Test
+  def testDownConversionWithConnectionFailure(): Unit = {
+    val (topicPartition, leaderId) = createTopics(numTopics = 1, numPartitions = 1).head
+
+    val producer = TestUtils.createNewProducer(TestUtils.getBrokerListStrFromServers(servers),
+      retries = 5, keySerializer = new StringSerializer, valueSerializer = new ByteArraySerializer)
+    try {
+      val bytes = new Array[Byte](100 * 1000)
+      for (i <- 0 to 3000) {
+        producer.send(new ProducerRecord(topicPartition.topic, topicPartition.partition,
+          "key", bytes))
+      }
+      producer.flush()
+    } finally {
+      producer.close()
+    }
+
+    def fetch(version: Short, waitForResponse: Boolean): Option[FetchResponse] = {
```

Comment:
Maybe `waitForResponse` should be `readPartialResponse` or something that indicates that we want to fail the request while the response is being read. Worth a brief comment too probably.

Reply:
Changed to `closeAfterPartialResponse`